### PR TITLE
DAOS-6593 control: updates to ras event handlers

### PR DESCRIPTION
**Pull request by tanabarr**
- Log events on the control-plane instance that manages the I/O
  Server that raised the event but don't log forwarded events at
  destination.
- Make format of the logged event consistent across data and control
  planes.
- Address other minor updates requested in PR-4394 code review.

Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>

### CODE REVIEW

File: src/control/events/ras.go [hunk 1]
```diff
@@ -127,6 +129,20 @@ type RASEvent struct {
...
+// IsForwarded returns true if event has been forwarded between hosts.
+func (evt *RASEvent) IsForwarded() bool {
+	return evt.forwarded
+}
+
+// WithIsForwarded sets the forwarded state of this event.
+func (evt *RASEvent) WithIsForwarded(isForwarded bool) *RASEvent {
+	evt.forwarded = isForwarded
+
+	return evt
```
Comment (mjmac) [inline src/control/events/ras.go:145]: As long as there is no read of evt.forwarded before the write (and then of course there are no subsequent writes), I think that this should be safe to use concurrently with no synchronization. I would have to take a closer look to see if that's the case, though.

File: src/control/lib/control/system.go [hunk 2]
```diff
@@ -248,7 +248,7 @@ func (fwdr *EventForwarder) OnEvent(ctx context.Context, evt *events.RASEvent) {
 
 	req := &SystemNotifyReq{
 		Sequence: <-fwdr.seq,
-		Event:    evt,
+		Event:    evt.WithIsForwarded(true),
```
Comment (mjmac) [inline src/control/lib/control/system.go:251]: I'm not sure how this works... Isn't this helper setting an unexported field? If so, the value won't survive the trip across gRPC, unless I'm missing something.

File: src/control/events/ras.go [hunk 3]
```diff
@@ -222,6 +238,57 @@ func (evt *RASEvent) FromProto(pbEvt *sharedpb.RASEvent) (err error) {
 	return
 }
 
+// PrintRAS generates a string representation of the event consistent with
+// what is logged in the data plane.
+func (evt *RASEvent) PrintRAS() string {
```
Comment (mjmac) [inline src/control/events/ras.go:243]: I wonder if it makes sense to put this helper in the `dmg/pretty` package. I don't feel too strongly either way, though.

Comment (kjacque) [reply src/control/events/ras.go:243]: I'd agree with putting it in dmg/pretty as long as it's only intended to be used in dmg.

File: src/control/lib/control/system.go (hunk 2 above)
Comment (kjacque) [reply src/control/lib/control/system.go:251]: The protobuf definition will need to be updated as well, won't it?

**Review by kjacque — COMMENTED** [review]
Didn't noticed anything further beyond @mjmac's comments. Is there any way to unit test the new forwarding logic?

File: src/control/events/ras.go (hunk 1 above)
Comment (tanabarr) [reply src/control/events/ras.go:145]: addressed by changing forwarded to an atomic bool, see https://github.com/daos-stack/daos/pull/4435

File: src/control/lib/control/system.go (hunk 2 above)
Comment (tanabarr) [reply src/control/lib/control/system.go:251]: fixed implementation in https://github.com/daos-stack/daos/pull/4435 , forwarded only exists in the native type and gets set on receipt in the gRPC ClusterEvent handler https://github.com/daos-stack/daos/pull/4435/files#diff-082855a22f45199564e850828dfc080b0d4c1b0dfa128256011737ccff1c1123R744 https://github.com/daos-stack/daos/pull/4435/files#diff-6e63051d5e46ac18fba5b9ba7508beda791d44a7dda83cebf0b2a3b95764eacbR299 , hence why it's un-exported

File: src/control/events/ras.go (hunk 3 above)
Comment (tanabarr) [reply src/control/events/ras.go:243]: it's used to log the messages and will eventually provide a structured output to syslog so I would prefer to leave it here for the moment (it's not used in dmg)

**Comment by tanabarr** [comment]
> Didn't noticed anything further beyond @mjmac's comments. Is there any way to unit test the new forwarding logic?

Unit testing of the forwarding and logging logic for the events is covered in the following tests (in the updated https://github.com/daos-stack/daos/pull/4435 PR):
lib/control/system_test.go TestControl_EventLogger_OnEvent & TestControl_EventForwarder_OnEvent
server/mgmt_system_test.go TestServer_MgmtSvc_ClusterEvent

**Comment by tanabarr** [comment]
I've merged this PR with the requirements for adding controlled shutdown and restart RAS events. the new PR is https://github.com/daos-stack/daos/pull/4435 and I've attempted to address comments from @kjacque and @mjmac there.

**Pull request closed without merging**